### PR TITLE
very basic mvp for shedding load

### PR DESCRIPTION
**Pull request by henryoswald**
### Description
I decided to try and keep load shedding as simple as possible for the first attempt. This is the simplest possible implementation I think we can do and should prove a useful test. We can initially deploy it to 1 instance group and watch how it behaves over a few days. 

The change made is when the box is over 100% capacity start to shed load.

If this proves successful we can then roll it out to more instance groups and possibly make shedding more aggressive (targeting ~90% cpu usage etc)

#### Related Issues / PRs
https://github.com/overleaf/issues/issues/4137

#### Potential Impact
- Medium, projects could stampeded all over our clsi's if we shed too much

### Deployment

- [ ] rake deploy:clsi[staging,master,latest]
- [ ] only deploy to 1 instance group and then cancel the deploy for a while

#### Metrics and Monitoring
https://console.cloud.google.com/monitoring/dashboards/builder/6376263191972157537?authuser=1&project=mgcp-1117973-ol-prod&dashboardBuilderState=%257B%2522editModeEnabled%2522:false%257D&timeDomain=1h

#### Who Needs to Know?
@emcsween @briangough @das7pad

### CODE REVIEW

**Review by gh2k — APPROVED** [review]
I'm not overly familiar with how this stuff works, but it matches the thing that Brian was talking about recently, so let's try it.

File: app.js [hunk 1]
```diff
@@ -319,7 +319,7 @@ const loadTcpServer = net.createServer(function (socket) {
 
   if (STATE === 'up' && Settings.internal.load_balancer_agent.report_load) {
     let availableWorkingCpus
-    const currentLoad = os.loadavg()[0]
+    const currentLoad = os.loadavg()[1]
```
Comment (das7pad) [inline app.js:322]: I suspect that this change will cause even more clsi nodes to get overloaded.

New projects should not get assigned to a VM already running 16+ compiles.

File: app.js [hunk 2]
```diff
@@ -331,7 +331,7 @@ const loadTcpServer = net.createServer(function (socket) {
     const freeLoad = availableWorkingCpus - currentLoad
     let freeLoadPercentage = Math.round((freeLoad / availableWorkingCpus) * 100)
     if (freeLoadPercentage <= 0) {
-      freeLoadPercentage = 1 // when its 0 the server is set to drain and will move projects to different servers
+      freeLoadPercentage = 0 // when its 0 the server is set to drain and will move projects to different servers
```
Comment (das7pad) [inline app.js:334]: WDYT about keeping the `currentLoad` as is and adjust the edge case to
```suggestion
      // when its 0 the server is set to drain and will move projects to different servers
      freeLoadPercentage = os.loadavg()[1] > availableWorkingCpus ? 0 : 1
```
(Maybe store `os.loadavg()` in a variable to save a sys call under heavy load.)

File: app.js (hunk 1 above)
Comment (henryoswald) [reply app.js:322]: that's fair, lets change 1 thing at a time.

File: app.js (hunk 2 above)
Comment (henryoswald) [reply app.js:334]: I'm missing the advantage of this. Wouldn't it mean we only have 2 values returned, 0 & 1?

**Comment by briangough** [comment]
I checked the current state of the clsi's and this will put 21 out of 74 instances into the "drain" state (28%) using the 1 minute load average (the 5 minute load average is essentially the same).  This seems fine to me for a trial. There's plenty of spare capacity on the other instances. 

A metric that would be useful here is the average time between a forced change of server.

File: app.js (hunk 2 above)
Comment (das7pad) [reply app.js:334]: This is the edge case handling for overloaded CPUs (load1 of 50 for a 5 core box). The proposed change would keep overloaded instances in rotation (with weight 1) for about 5min until the load5 hits the capacity limit.

On a 16 core box:

time in s | load1 | load5 | weight
--- | --- | --- | ---
0 | 0 | 0 | 100
30 | 5 | 0 | 66
60 | 16 | 5 | 1
300 | 16 | 16 | 0

Comment (das7pad) [reply app.js:334]: (Delaying the drain mode (weight=0) for 5min might stop a stamped of projects in a bad state.)

Comment (henryoswald) [reply app.js:334]: Right, so you were suggesting load5 for going to 0 capacity and load1 for looking moving us to 1. It's a good idea, for the moment I think I want to keep it a bit simpler and just use the single load value, in the next iteration I might look at this. Cheers.

**Review by das7pad — APPROVED** [review]

**Comment by henryoswald** [comment]
> A metric that would be useful here is the average time between a forced change of server.

so we know the rate of projects being moved off? e.g. 30 projects/min.

**Comment by briangough** [comment]
> > A metric that would be useful here is the average time between a forced change of server.
> 
> so we know the rate of projects being moved off? e.g. 30 projects/min.

Yes, we could  get that information in web from the ClsiCookie module if we recorded the time the clsi server entry in redis was first set, and then use that when we see it change after a compile.

**Pull request merged**